src/Agent.Worker/AgentLogPlugin.cs

```diff
@@ -0,0 +1,228 @@
+
+                // construct plugin context
+                AgentLogPluginHostContext pluginContext = new AgentLogPluginHostContext
+                {
+                    PluginAssemblies = new List<string>(),
+                    Repositories = context.Repositories,
+                    Endpoints = context.Endpoints,
+                    Variables = new Dictionary<string, VariableValue>(),
+                    Steps = new Dictionary<string, Pipelines.TaskStepDefinitionReference>()
+                };
+
+                // plugins 
+                pluginContext.PluginAssemblies.AddRange(_logPlugins.Values.Select(x => x.AssemblyName));
+
+                // variables
+                foreach (var publicVar in context.Variables.Public)
+                {
+                    pluginContext.Variables[publicVar.Key] = publicVar.Value;
+                }
+                foreach (var publicVar in context.Variables.Private)
+                {
+                    pluginContext.Variables[publicVar.Key] = new VariableValue(publicVar.Value, true);
+                }
+
+                // steps
+                foreach (var step in steps)
+                {
+                    var taskStep = step as ITaskRunner;
+                    if (taskStep != null)
+                    {
+                        pluginContext.Steps[taskStep.ExecutionContext.Id.ToString("D")] = taskStep.Task.Reference;
+                    }
+                }
+
+                Trace.Info("Send serialized context through STDIN");
+                _redirectedStdin.Enqueue(JsonUtility.ToString(pluginContext));
+
+                foreach (var plugin in _logPlugins)
+                {
+                    context.Output($"Plugin: '{plugin.Value.FriendlyName}' is running in background.");
+                }
+            }
+
+            return Task.CompletedTask;
+        }
+
+        public async Task WaitAsync(IExecutionContext context)
+        {
+            Trace.Entering();
+
+            if (_pluginHostProcess != null)
+            {
+                Trace.Info("Send instruction code through STDIN to stop plugin host");
+
+                // plugin host will stop the routine process and give every plugin a chance to participate into job finalization
+                _redirectedStdin.Enqueue($"##vso[logplugin.finish]{_instanceId.ToString("D")}");
+
+                // print out outputs from plugin host and wait for plugin finish
+                Trace.Info("Waiting for plugin host exit");
+                foreach (var plugin in _logPlugins)
+                {
+                    context.Debug($"Waiting for log plugin '{plugin.Value.FriendlyName}' to finish.");
+                }
+
+                while (!_pluginHostProcess.IsCompleted)
+                {
+                    while (_outputs.TryDequeue(out string output))
+                    {
+                        if (output.StartsWith("##[plugin.trace]", StringComparison.OrdinalIgnoreCase))
```

Comment:
Consider either make `##[plugin.trace]` a shared constant, or extract it into a local variable.

Reply:
sure